### PR TITLE
Don't use for-in loop in Interpolator.sort()

### LINKED ISSUE #10320: Plotting with transforms behavior change
An application with a selected-point, constrained to move along a reference curve via interactive user control, is implemented using bokeh transforms. This works as expected in bokeh 2.0.2. In bokeh 2.1.0 (and 2.1.1), the web page excludes the associated plots, and errors are seen in the Javascript console of the client browser.

See bokeh discourse topic [Plotting changes between 2.0.2 and 2.1.x](https://discourse.bokeh.org/t/bokeh-plotting-changes-between-2-0-2-and-2-1-x/6004/3) for additional details.

A minimal reproducible example follows, which can highlight the problem with simple plot/glyph creation, excluding any interactive elements.

```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
"""
import numpy as np

from bokeh.plotting import figure, show
from bokeh.transform import transform

from bokeh.models import ColumnDataSource
from bokeh.models import LinearInterpolator

# Referencce curve
x = np.linspace(0.0,1.0,1001)
y = np.tanh(np.pi*x)

Sref = ColumnDataSource(data=dict(x=x, y=y))

# Operating point constrained to reference curve
op = 301
Ssel = ColumnDataSource(data=dict(x=[x[op]], y=y[[op]]))

_li_x = LinearInterpolator(x='x', y='x', data=Sref, clip=False)
_js_x = transform('x', _li_x)
_li_y = LinearInterpolator(x='x', y='y', data=Sref, clip=False)
_js_y = transform('x', _li_y)


p = figure(width=400, height=300)
p.line(x='x', y='y', source=Sref)
p.inverted_triangle(x=_js_x, y=_js_y, source=Ssel, size=15, line_color='#000000', fill_color='#000000')

show(p)
```

### PR DESCRIPTION
**Pull request by mattpap**
- [x] tests

fixes #10320

### CODE REVIEW

**Comment by bryevdv** [comment]
OP reports things working in 2.0.2 so what actually changed?

File: bokehjs/src/lib/models/transforms/interpolator.ts [hunk 1]
```diff
@@ -83,8 +83,8 @@ export abstract class Interpolator extends Transform {
     // The following sorting code is referenced from:
     // http://stackoverflow.com/questions/11499268/sort-two-arrays-the-same-way
     const list: {x: number, y: number}[] = []
-    for (const j in tsx) {
-      list.push({x: tsx[j], y: tsy[j]})
+    for (let i = 0; i < tsx.length; i++) {
```
Comment (p-himik) [inline bokehjs/src/lib/models/transforms/interpolator.ts:86]: I would definitely put a comment here explaining why a simpler `const j in tsx` cannot be used.

**Comment by p-himik** [comment]
@bryevdv https://github.com/bokeh/bokeh/pull/10069 was implemented that introduced some new fields. `const j in ...` started to enumerate those fields along with the numerical indices.

**Comment by mattpap** [comment]
Expanding on @p-himik's comment, using `for-in` loop to obtain indices of an array, is a historical artifact and something one shouldn't do. It works with built-in arrays, because they were designed to allow this, at the cost added complexity and lack of (or tedious) extensibility. I'm a bit surprised this works with typed arrays. This is definitively not going to work with derived types, unless they specifically designed to allow this. I'm in the process of removing `for-in` loops from bokehjs' code base, and I'm going to disallow them altogether at some point. The only acceptable ways to iterate an array is to either use explicit indexing (if an index is needed), iterators (in particular using `for-of` loop, because that can be more efficient (e.g. in FF; actually a lot) than `for(;;)` loop)) or `.forEach()` (at al) (which I don't recommend with large arrays as their performance is usually pretty bad for some reason; use `core/util/array*` instead as they have well defined performance characteristics). Use `keys()`, `values()` and `entries()` (from `core/util/object`) to iterate over plain objects.

File: bokehjs/src/lib/models/transforms/interpolator.ts (hunk 1 above)
Comment (mattpap) [reply bokehjs/src/lib/models/transforms/interpolator.ts:86]: I disagree with an inline comment, but I would add it to a "good practices" wiki page or dev docs. However, as I mentioned in the main comment, I will disallow "unauthorized" `for-in` loops altogether in near future (at jslint level).

**Comment by p-himik** [comment]
> I'm going to disallow them altogether at some point

Is it possible to disallow `for-in` loops only for types that are not built-in? Because, given the better performance of `for-of`, one could expect a similar benefit for `for-in`. Also, it's still simpler than the explicit version with an increment and a comparison.

> This is definitively not going to work with derived types, unless they specifically designed to allow this.

Given the above, wouldn't it be better to just design the derived types in such a way so that `for-in` loops are supported? Especially if it's possible to disallow using `for-in` for all other types.

**Comment by mattpap** [comment]
> Also, it's still simpler than the explicit version with an increment and a comparison.

Syntactically sure, but semantically not really:
```ts
> x = new Float64Array([0,1,2])
Float64Array(3) [ 0, 1, 2 ]
> y = []
[]
> for (const i in x) y.push(i)
3
> y.map((i) => typeof i)
[ 'string', 'string', 'string' ]
```
I would expect `'number'` type instead, if this was to be relied on. `for-in` loops (and `in` operator for that that matter) are a nightmare under refactoring even in a statically typed codebase. This code also assumes that the input is a "well behaved" array type, but that's not what the interface says, as `tsx` is `Arrayable`, which doesn't provide `[key: string]: T` interface (but that's the hard part of statically typing `for-in` and `in`).

My honest advice is to not use such quirks in general and this one in particular, as it is simply not worth the trouble. Even if this potential micro-optimisation was an actual optimisation (I doubt that), it's being canceled out many times over by the overall inefficient algorithm it's a part of. If GC could speak, it wouldn't be polite about this code, i.e. creating an object per array item.

> (...) couldn't it be better to just design the derived types in such a way so that for-in loops are supported?

This is a problem when old meets new. We can either uphold the obsolete interface of built-in arrays or have decent ES2015 classes (with enumerable properties). Given my conclusions regarding `for-in` loop with arrays, I think that upholding the interface shouldn't be a concern. However, all this discussion may be mute, because this current approach doesn't work with legacy bundles, and I may need to roll it back.

**Pull request merged**